blog: Running at Scale (#80)
eclipse-che/blog: _posts/2025-04-28-running-at-scale.adoc

While there is no strict limit on the number of resources in a Kubernetes cluster, 
there are certain thresholds to keep in mind. For example, link:https://www.redhat.com/en/technologies/cloud-computing/openshift[OpenShift Container Platform], which is a certified distribution of Kubernetes, provides a set of tested maximums for various resources, which can serve as an initial guideline for planning your environment:

[%header,format=csv]
|===
Resource type, Tested maximum
Number of nodes,2000
Number of pods,150000
Number of pods per node,2500
Number of namespace,10000
Number of services,10000
Number of secrets,80000
Number of config maps,90000
|===

Table 1: OpenShift Container Platform tested cluster maximums for various resources.

*TODO: Is there anything similar for vanilla Kubernetes ?*

TIP: More details on the OpenShift Container Platform tested object maximums can be found in the link:https://docs.redhat.com/en/documentation/openshift_container_platform/4.18/html/scalability_and_performance/planning-your-environment-according-to-object-maximums#planning-your-environment-according-to-object-maximums[official documentation].

> For example, having over 10,000 is generally not recommended due to potential performance and management overhead. In Eclipse Che, for instance, each user is allocated a namespace. If you expect the user base to be large, consider spreading workloads across multiple "fit-for-purpose" clusters and potentially leveraging solutions for multi-cluster orchestration.

Reviewer: ```suggestion
For example, having over 10,000 namespaces is generally not recommended due to potential performance and management overhead. In Eclipse Che, for instance, each user is allocated a namespace. If you expect the user base to be large, consider spreading workloads across multiple "fit-for-purpose" clusters and potentially leveraging solutions for multi-cluster orchestration.
```
?

=== Calculating Resource Requirements

> When deploying Eclipse Che on Kubernetes, it is crucial to accurately calculate the resource requirements and determine memory and CPU / GPU needs for each Cloud Development Environment (CDE) to come up with the right sizing of the cluster. In general, the CDE size is limited by the worker node size and can not be bigger than it. The resource requirements for CDEs can vary significantly based on the specific workloads and configurations. For example, a trivial CDE may require only a few hundred megabytes of memory, while a more complex one will need several gigabytes of memory and multiple CPU cores.

Author: @deerskindoll thank you for the review, this block seems to be outdated, could you please re-read and provide updated suggestion if needed

IMPORTANT: The default etcd size is 2 GB, and the recommended maximum is 8 GB. Exceeding the maximum limit can make the Kubernetes cluster unstable and unresponsive.

image::/assets/img/running-at-scale/etcd.png[The metrics showing the progression of the etcd storage growth]

Figure 1: Metrics with the progression of the etcd storage growth that leads to cluster instability.

=== The size of the objects matters

Not only the overall number, but also the size of the objects stored in etcd is a critical factor that can significantly impact its performance and stability. Each object stored in etcd consumes space, and as the number of objects increases, the overall size of etcd grows. The larger the object is, the more space it takes in etcd. For example, etcd can be overloaded with just a few thousand of relatively big Kubernetes objects.

IMPORTANT: Even though the data stored in a `ConfigMap` cannot exceed 1 MiB by design, a few thousand of relatively big `ConfigMap` objects can overload etcd storage.

> In the context of Eclipse Che, the operator has a feature that merges the cluster's Certificate Authorities (CAs) bundle into a `ConfigMap` and provisions it for each user namespace, resulting in additional etcd usage, given that there is a large number of TLS certificates in the cluster.

Reviewer 2: ```suggestion
In the context of Eclipse Che, the operator has a feature that merges the cluster's Certificate Authorities (CAs) bundle into a `ConfigMap` and provisions it for each user namespace. This results in additional etcd usage, given that there is a large number of TLS certificates in the cluster.
```

> In order to disable mounting of the CA bundle using the `ConfigMap` under the path `/etc/pki/ca-trust/extracted/pem`, configure the `CheCluster` Custom Resource by setting `disableWorkspaceCaBundleMount` property to `true`. Certificates will be mounted under the path `/public-certs` in this case:

Reviewer 2: ```suggestion
In order to disable mounting the CA bundle using the `ConfigMap` under the `/etc/pki/ca-trust/extracted/pem` path, configure the `CheCluster` Custom Resource by setting the `disableWorkspaceCaBundleMount` property to `true`. Certificates will be mounted under the path `/public-certs` in this case:
```

> In order to disable mounting the CA bundle using the `ConfigMap` under the `/etc/pki/ca-trust/extracted/pem` path, configure the `CheCluster` Custom Resource by setting the `disableWorkspaceCaBundleMount` property to `true`. Certificates will be mounted under the path `/public-certs` in this case:

Reviewer 3: ```suggestion
In order to disable mounting the CA bundle using the `ConfigMap` under the `/etc/pki/ca-trust/extracted/pem` path, configure the `CheCluster` Custom Resource by setting the `disableWorkspaceCaBundleMount` property to `true`. With this configuration, certificates will be mounted under the path `/public-certs`:
```

```yaml
spec:
  devEnvironments:
    trustedCerts:
      disableWorkspaceCaBundleMount: true
```

> TIP: More details aboutimporting untrusted TLS certificates can be found in the link:https://eclipse.dev/che/docs/stable/administration-guide/importing-untrusted-tls-certificates/[official documentation].

Reviewer 4: ```suggestion
TIP: More details about importing untrusted TLS certificates can be found in the link:https://eclipse.dev/che/docs/stable/administration-guide/importing-untrusted-tls-certificates/[official documentation].
```

=== DevWorkspace objects

For large Kubernetes deployments, particularly those involving a high number of custom resources such as `DevWorkspace` objects, which represent Cloud Development Environments, etcd can become a significant performance bottleneck. 

> IMPORTANT: Based on the load testing 6,000 of `DevWorkspace` objects consume around 2.5GB of the etcd storage.

Reviewer 2: ```suggestion
IMPORTANT: Based on the load testing for 6,000 `DevWorkspace` objects, storage consumption for etcd was around 2.5GB.
```

> Starting from link:https://github.com/devfile/devworkspace-operator[DevWorkspace Operator] 0.34.0 version, it is possible to configure the pruner that would automatically clean up the `DevWorkspace` objects that are not in use for a certain period of time. This can be done by configuring the `DevWorkspaceOperatorConfig` object accordingly: 

Reviewer 2: ```suggestion
Starting from link:https://github.com/devfile/devworkspace-operator[DevWorkspace Operator] version 0.34.0, it is possible to configure a pruner that would automatically clean up `DevWorkspace` objects that were not in use for a certain period of time. This can be done by configuring the `DevWorkspaceOperatorConfig` object accordingly: 
```

> Starting from link:https://github.com/devfile/devworkspace-operator[DevWorkspace Operator] version 0.34.0, it is possible to configure a pruner that would automatically clean up `DevWorkspace` objects that were not in use for a certain period of time. This can be done by configuring the `DevWorkspaceOperatorConfig` object accordingly: 

Reviewer 3: ```suggestion
Starting from link:https://github.com/devfile/devworkspace-operator[DevWorkspace Operator] version 0.34.0, you can configure a pruner that would automatically clean up `DevWorkspace` objects that were not in use for a certain period of time. To set the pruner up, configure the `DevWorkspaceOperatorConfig` object as follows: 
```

```yaml
apiVersion: controller.devfile.io/v1alpha1
kind: DevWorkspaceOperatorConfig
metadata:
  name: devworkspace-operator-config
  namespace: crw
config:
  workspace:
    cleanupCronJob:
      enabled: true
      dryRun: false
      retainTime: "60"
      schedule: "* * * * *"
```

> TIP: More details about DevWorkspace Operator Configuration can be found in the link:https://github.com/devfile/devworkspace-operator/blob/main/docs/dwo-configuration.md[official documentation].

Reviewer 3: ```suggestion
TIP: You can find more details about DevWorkspace Operator Configuration in the link:https://github.com/devfile/devworkspace-operator/blob/main/docs/dwo-configuration.md[official documentation].
```

*TODO: merge https://github.com/devfile/devworkspace-operator/pull/1402 and update docs*

=== OLMConfig

> When an operator is installed by the link:https://olm.operatorframework.io/[Operator Lifecycle Manager (OLM)], a stripped-down copy of its CSV is created in every namespace the operator is configured to watch. These stripped-down CSVs are known as “Copied CSVs” and communicate to users which controllers are actively reconciling resource events in a given namespace. On especially large clusters, with namespaces and installed operators tending in the hundreds or thousands, Copied CSVs consume an untenable amount of resources; e.g. OLM’s memory usage, cluster etcd limits, networking, etc. In order to eliminate the CSVs copied to every namespace, the `OLMConfig` object can be configured accordingly:

Reviewer 3: ```suggestion
When an operator is installed by the link:https://olm.operatorframework.io/[Operator Lifecycle Manager (OLM)], a stripped-down copy of its CSV is created in every namespace the operator is configured to watch. These stripped-down CSVs are known as “Copied CSVs” and communicate to users which controllers are actively reconciling resource events in a given namespace. On especially large clusters, with namespaces and installed operators tending in the hundreds or thousands, Copied CSVs consume an untenable amount of resources; e.g. OLM’s memory usage, cluster etcd limits, networking, etc. In order to eliminate the CSVs copied to every namespace, configure the `OLMConfig` object accordingly:
```

```yaml
apiVersion: operators.coreos.com/v1
kind: OLMConfig
metadata:
  name: cluster
spec:
  features:
    disableCopiedCSVs: true
```

The primary impact of the `disableCopiedCSVs` property on etcd is related to resource consumption. In clusters with a large number of namespaces and many cluster-wide Operators, the creation and maintenance of numerous Copied CSVs can lead to increased etcd storage usage and memory consumption. By disabling Copied CSVs, the amount of data stored in etcd is significantly reduced, which can help improve overall cluster performance and stability.

> This is particularly important in large clusters where the number of namespaces and operators can quickly add up to a significant amount of data. Disabling Copied CSVs can help reduce the load on etcd, leading to improved performance and responsiveness of the cluster.

Reviewer 2: ```suggestion
This is particularly important for large clusters where the number of namespaces and operators can quickly add up to a significant amount of data. Disabling Copied CSVs can help reduce the load on etcd, leading to improved performance and responsiveness of the cluster.
```

Additionally, it can help reduce the memory footprint of OLM, as it no longer needs to maintain and manage these additional resources.

> TIP: More details about Disabling Copied CSVs can be found in the link:https://olm.operatorframework.io/docs/advanced-tasks/configuring-olm/#disabling-copied-csvs[official documentation].

Reviewer 3: ```suggestion
TIP: You can find more details about Disabling Copied CSVs in the link:https://olm.operatorframework.io/docs/advanced-tasks/configuring-olm/#disabling-copied-csvs[official documentation].
```

== Cluster Autoscaling

> Although cluster autoscaling is a powerful Kubernetes feature, you cannot always fall back on it and should always consider predictive scaling by analyzing load data on your environment to detect daily or weekly usage patterns. If your workloads follow a pattern and there are dramatic peaks throughout the day, you should consider provisioning worker nodes accordingly. For example, if you have a predictable load pattern where the number of workspaces increases during business hours and decreases during off-hours, you can use predictive scaling to adjust the number of worker nodes based on the expected load.

Reviewer 3: ```suggestion
Although cluster autoscaling is a powerful Kubernetes feature, you cannot always fall back on it. You should always consider predictive scaling by analyzing load data on your environment to detect daily or weekly usage patterns. If your workloads follow a pattern and there are dramatic peaks throughout the day, you should consider provisioning worker nodes accordingly. For example, if you have a predictable load pattern where the number of workspaces increases during business hours and decreases during off-hours, you can use predictive scaling to adjust the number of worker nodes based on the expected load.
```

This can help ensure that you have enough resources available to handle the peak load while minimizing costs during off-peak hours.

== Multicluster

> By design, Eclipse Che is not multicluster aware, and you can only have one instance per cluster. However, it is possible to run Eclipse Che in a multicluster environment by deploying Eclipse Che in different clusters and using a load balancer or DNS-based routing to direct traffic to the appropriate instance based on the user’s location or other criteria. This approach can help improve performance and reliability by distributing the workload across multiple clusters and providing redundancy in case of cluster failures.

Reviewer 2: ```suggestion
By design, Eclipse Che is not multicluster aware, and you can only have one instance per cluster. However, it is possible to run Eclipse Che in a multicluster environment by deploying Eclipse Che in each different cluster and using a load balancer or DNS-based routing to direct traffic to the appropriate instance based on the user’s location or other criteria. This approach can help improve performance and reliability by distributing the workload across multiple clusters and providing redundancy in case of cluster failures.
```

> By design, Eclipse Che is not multicluster aware, and you can only have one instance per cluster. However, it is possible to run Eclipse Che in a multicluster environment by deploying Eclipse Che in each different cluster and using a load balancer or DNS-based routing to direct traffic to the appropriate instance based on the user’s location or other criteria. This approach can help improve performance and reliability by distributing the workload across multiple clusters and providing redundancy in case of cluster failures.

Reviewer 3: ```suggestion
By design, Eclipse Che is not multicluster aware, and you can only have one instance per cluster. However, you can run Eclipse Che in a multicluster environment by deploying Eclipse Che in each different cluster and using a load balancer or DNS-based routing to direct traffic to the appropriate instance based on the user’s location or other criteria. This approach can help improve performance and reliability by distributing the workload across multiple clusters and providing redundancy in case of cluster failures.
```

One example of a multicluster architecture is link:https://workspaces.openshift.com/[workspaces.openshift.com], which is part of the link:https://developers.redhat.com/developer-sandbox[Red Hat Developer Sandbox].

From the infrastructure perspective, the Red Hat Developer Sandbox consists of multiple link:https://www.redhat.com/en/technologies/cloud-computing/openshift/aws[ROSA] clusters. On each cluster, the productized version of Eclipse Che is installed and configured using link:https://argo-cd.readthedocs.io/en/stable/[Argo CD]. Since the user base is spread across multiple clusters, link:https://workspaces.openshift.com/[workspaces.openshift.com] is used as a single entry point to the productized Eclipse Che instances. You can find implementation details about the multicluster redirector in the following link:https://github.com/codeready-toolchain/crw-multicluster-redirector[GitHub repository].

> TIP: link:https://workspaces.openshift.com/[workspaces.openshift.com] is a Developer Sandbox-specific solution that can not be reused as-is in other environments. However, it can be used as a reference for implementing a similar solution well-tailored for your specific multicluster needs.

Reviewer 2: ```suggestion
TIP: The solution for link:https://workspaces.openshift.com/[workspaces.openshift.com] is a Developer Sandbox-specific solution that can not be reused as-is in other environments. However, it can be used as a reference for implementing a similar solution well-tailored for your specific multicluster needs.
```

> TIP: The solution for link:https://workspaces.openshift.com/[workspaces.openshift.com] is a Developer Sandbox-specific solution that can not be reused as-is in other environments. However, it can be used as a reference for implementing a similar solution well-tailored for your specific multicluster needs.

Reviewer 3: ```suggestion
TIP: The solution for link:https://workspaces.openshift.com/[workspaces.openshift.com] is a Developer Sandbox-specific solution that can not be reused as-is in other environments. However, you can use it as a reference for implementing a similar solution well-tailored to your specific multicluster needs.
```

TODO: review https://issues.redhat.com/browse/CRW-8595
 
== Conclusion

> Running Eclipse Che at scale on Kubernetes presents unique challenges that require careful planning and consideration of various factors. By understanding the limitations of Kubernetes, accurately calculating resource requirements, and implementing best practices for managing etcd and OLM, organizations can build a robust and scalable CDE platform that meets the needs of their users. Additionally, leveraging predictive scaling and considering multicluster architectures can further enhance the performance and reliability of Eclipse Che deployments. By following these guidelines and continuously monitoring and optimizing the environment, organizations can provide a reliable and efficient Cloud Development Environment, ensure a seamless and efficient development experience for a large number of users while maintaining responsiveness, and prevent performance degradation of the cluster.

Reviewer 3: ```suggestion
Running Eclipse Che at scale on Kubernetes presents unique challenges that require careful planning and consideration of various factors. By understanding the limitations of Kubernetes, accurately calculating resource requirements, and implementing best practices for managing etcd and OLM, you can build a robust and scalable CDE platform that meets the needs of your users. Additionally, leveraging predictive scaling and considering multicluster architectures can further enhance the performance and reliability of Eclipse Che deployments. By following these guidelines and continuously monitoring and optimizing the environment, you can provide a reliable and efficient CDE, ensure a seamless and efficient development experience for a large number of users while maintaining responsiveness, and prevent performance degradation of the cluster.
```

[%header,format=csv]
|===
Resource type, Tested maximum
Number of nodes,2000
Number of pods,150000
Number of pods per node,2500
Number of namespace,10000
Number of services,10000
Number of secrets,80000
Number of config maps,90000
|===

Table 1: OpenShift Container Platform tested cluster maximums for various resources.

TIP: You can find more details on the OpenShift Container Platform tested object maximums in the link:https://docs.redhat.com/en/documentation/openshift_container_platform/4.18/html/scalability_and_performance/planning-your-environment-according-to-object-maximums#planning-your-environment-according-to-object-maximums[official documentation].

For example, it is generally not recommended to have more than 10,000 namespaces due to potential performance and management overhead. In Eclipse Che, for instance, each user is allocated a namespace. If you expect the user base to be large, consider spreading workloads across multiple "fit-for-purpose" clusters and potentially leveraging solutions for multi-cluster orchestration.

=== Calculating Resource Requirements

> When deploying Eclipse Che on Kubernetes, it is crucial to accurately calculate the resource requirements and determine memory and CPU / GPU needs for each Cloud Development Environment (CDE) to come up with the right sizing of the cluster. In general, the CDE size is limited by the worker node size and can not be bigger than that. The resource requirements for CDEs can vary significantly based on the specific workloads and configurations. For example, a trivial CDE may require only a few hundred megabytes of memory, while a more complex one will need several gigabytes of memory and multiple CPU cores.

Reviewer 3: ```suggestion
When deploying Eclipse Che on Kubernetes, it is crucial to accurately calculate the resource requirements and determine memory and CPU / GPU needs for each CDE to come up with the right sizing of the cluster. In general, the CDE size is limited by and can not be bigger than the worker node size. The resource requirements for CDEs can vary significantly based on the specific workloads and configurations. For example, a simple CDE may require only a few hundred megabytes of memory, while a more complex one will need several gigabytes of memory and multiple CPU cores.
```

Etcd is the primary datastore of Kubernetes clusters configuration and state. In particular, applications on the cluster and the cluster itself both generate resources (data) that are stored in etcd.  As a distributed key-value store, etcd doesn’t scale well past a certain threshold, and as the size of etcd grows, so does the load on the cluster, risking its stability. As a mitigation and following a common practice adopted in Kubernetes 8 GiB limit to etcd size.

image::/assets/img/running-at-scale/etcd.png[The metrics showing the progression of the etcd storage growth]

Figure 1: The metrics showing the progression of the etcd storage growth.

== DevWorkspace Operator Prunner

Cleanup stale workspaces automatically 

https://github.com/devfile/devworkspace-operator/pull/1402

=== OLMConfig

```yaml
apiVersion: operators.coreos.com/v1
kind: OLMConfig
metadata:
  name: cluster
spec:
  features:
    disableCopiedCSVs: true

> ```

Author: === Image Puller ?

---
title: "Running at Scale"
layout: post
author: Ilya Buziuk
description: >-
  Running Eclipse Che at scale on Kubernetes
categories: []
keywords: ['Kubernetes', 'CDE', 'Cloud', 'Scalability', 'Eclipse Che']
slug: /@ilya.buziuk/running-at-scale
---

== Introduction

> Cloud Development Environments (CDEs) represent a paradigm shift in software development, offering on-demand, pre-configured, and often ephemeral environments that streamline the coding, testing, and debugging processes. By providing developers with consistent and isolated workspaces, CDEs enhance productivity, improve collaboration, and reduce the complexities associated with managing local development setups. The increasing demand for rapid software delivery and the adoption of cloud-native architectures have made scalable CDEs a critical component of modern software development infrastructure. Kubernetes, an open-source container orchestration platform, has emerged as a powerful foundation for deploying and managing containerized applications at scale. Its capabilities in automating deployment, scaling, and recovery, coupled with its robust resource management features, make it well-suited for orchestrating the dynamic and resource-intensive nature of CDEs. Organizations are increasingly leveraging Kubernetes to build scalable CDE platforms that can adapt to the fluctuating demands of development teams. However, achieving scale with CDEs on Kubernetes, particularly in the range of thousands concurrent workspaces, presents significant challenges. Such a scale imposes substantial infrastructure demands and introduces potential bottlenecks that can impact the performance and stability of the entire system. Addressing these challenges requires meticulous planning, strategic architectural choices, and continuous optimization to ensure a seamless and efficient development experience for a large number of users. This report aims to delve into the potential bottlenecks that may arise when running CDEs at this scale on Kubernetes, explore the best practices for mitigating these issues, and provide insights into building a robust and scalable CDE platform. The analysis will cover key areas such as etcd performance, Operator Lifecycle Manager (OLM) scalability, resource management, network configuration, and security considerations, drawing upon research and real-world experiences in large-scale Kubernetes deployments. The sheer volume of concurrent workspaces implies a high degree of dynamism within the Kubernetes cluster, characterized by frequent creation and deletion of pods and a significant load on core components like the API server and etcd. This necessitates a proactive and optimized approach to infrastructure management to maintain responsiveness and prevent performance degradation.

Reviewer 2: This paragraph is quite long, could we break it up to 2-3 paragraphs?

> To optimize etcd performance at scale, several best practices should be followed. Using low-latency storage, such as SSDs or NVMe drives on dedicated disks, is crucial for fast read and write operations. Proper etcd cluster sizing, typically with an odd number of members like 3 or 5, helps maintain quorum and fault tolerance without excessive replication overhead. Ensuring that etcd nodes have sufficient CPU and memory resources is also essential for stable operation. Continuously monitoring etcd's health and performance by tracking key metrics such as latency, throughput, and database size allows for early detection of potential issues. Regular maintenance, including compaction and defragmentation of the etcd database, helps reclaim storage space and improve performance. In very large clusters with high pod churn, considering a separate dedicated etcd instance specifically for Kubernetes Event objects can help alleviate the load on the main etcd cluster. Optimizing the way etcd clients, such as the API server, use etcd's API by avoiding large values in Put calls and minimizing frequent updates can also contribute to better performance. Finally, tuning etcd's configuration parameters, such as the database quota (--quota-backend-bytes), heartbeat interval (--heartbeat-interval), and leader election timeout (--election-timeout), might be necessary to tailor its behavior to the specific workload. Given the anticipated high volume of custom resources for CDE workspaces, a proactive approach to managing etcd's data footprint is essential, including regular monitoring of etcd object sizes per type to prevent control plane instability.

Reviewer 2: > To optimize etcd performance at scale, several best practices should be followed.

There are a lot of different best practices mentioned in this paragraph, WDYT about formatting them in a bullet point list?

ex:
* Using low-latency storage, such as SSDs or NVMe drives on dedicated disks, is crucial for fast read and write operations.
* Proper etcd cluster sizing, typically with an odd number of members like 3 or 5, helps maintain quorum and fault tolerance without excessive replication overhead.
* Ensuring that etcd nodes have sufficient CPU and memory resources is also essential for stable operation.
* Continuously monitoring etcd’s health and performance by tracking key metrics such as latency, throughput, and database size allows for early detection of potential issues.
